server/src/storage/core/storage.core.service.ts

```diff
@@ -0,0 +1,101 @@
+import { FileExtensionEnum, FileUpload, StorageFileCore } from "./types.core";
+import { ProvidersEnum } from "../providers";
+import { LocalStorageService } from "../providers/local/local.storage.service";
+import { BadRequestException, StreamableFile } from "@nestjs/common";
+
+export interface ProviderService {
+  uploadFile(file: FileUpload): Promise<StorageFileCore>;
+  downloadFile(file: StorageFileCore): Promise<StreamableFile>;
+  deleteFile(file: StorageFileCore): Promise<boolean>;
```

Comment:
Defining this types as generics that extends `StorageFileCore` instead of just relying on mapping would make the solution more type safe here.
i.e.
```suggestion
  uploadFile<T extends StorageFileCore>(file: FileUpload): Promise<T>;
  downloadFile<T extends StorageFileCore>(file: T): Promise<StreamableFile>;
  deleteFile<T extends StorageFileCore>(file: T): Promise<boolean>;
```